### PR TITLE
Login page

### PR DESCRIPTION
**Pull request by mk5med**
This PR introduces a login screen for users and uses JWT tokens to hold user sessions, and a logout button on the dashboard. From the login screen error codes are handled and printed for the benefit of the user. This PR extends #5.

screenshot

### CODE REVIEW

**Review by 1whatleytay — APPROVED** [review]

File: src/components/Header.vue [hunk 1]
```diff
@@ -127,8 +136,11 @@ export default {
   methods: {
     async commit() {
       await axios.put('/api/commit')
-
       this.$refs.commit.show()
+    },
+    async logout() {
```
Comment (1whatleytay) [inline src/components/Header.vue]:
```suggestion
    logout() {
```

It's probably a smart idea to tell the backend to invalidate this token too.

```
axios.put('/api/logout', localStorage.getItem('auth:token'))
```

File: src/router/index.js [hunk 2]
```diff
@@ -60,4 +70,25 @@ const router = new VueRouter({
   routes
 })
 
+router.beforeEach((to, from, next) => {
+  if (!to.meta.noAuth) {
+    const token = localStorage.getItem('auth:token')
+    if (token) {
+      var data = JSON.parse(atob(token.split('.')[1]))
```
Comment (1whatleytay) [inline src/router/index.js]:
```suggestion
      const data = JSON.parse(atob(token.split('.')[1]))
```

File: src/router/index.js [hunk 3]
```diff
@@ -60,4 +70,25 @@ const router = new VueRouter({
...
+      // If the token is expired, remove the token and navigate back to login
+      if (new Date(data.validUntil) > new Date()){
+        axios.defaults.headers['Authorization'] = `Bearer ${token}`
+      } else {
+        localStorage.removeItem('auth:token')
+        next('/login')
```
Comment (1whatleytay) [inline src/router/index.js]: Missing a return here. It will call `next('/login')` then `next()` and bug out. It would be nice if we could merge the two identical `next('/login')` paths with some trickery.

File: src/views/Login.vue [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+      <div>
+        <small class="font-light text-gray-500 text-xs mb-2"> Username </small>
+        <input
+          type="text"
+          class="border rounded p-2 text-lg w-full bg-gray-100 font-mono"
+          v-model="username"
+          required
+        />
+        <small class="font-light text-gray-500 text-xs mb-2 mt-4">
```
Comment (1whatleytay) [inline src/views/Login.vue]: Looking at the image, I would increase top margin here.
```suggestion
        <small class="font-light text-gray-500 text-xs mb-2 mt-8">
```

File: src/views/Login.vue [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        <img
+          class="w-16 h-16 mr-4 rounded-full"
+          src="@/assets/tu20logo.png"
+          alt="TU20 Logo"
+        />
+        TU20 Bot Config
+      </div>
+
+      <div>
+        <small class="font-light text-gray-500 text-xs mb-2"> Username </small>
```
Comment (1whatleytay) [inline src/views/Login.vue:14]: I don't remember what your justification for using the `<small>` tag was 😓 

The tag itself is useless and behaves exactly the same as `<span>` because of normalize.css. It doesn't tell the programmer any more about the component than `text-xs` does. I don't mind you keeping it here, but it's definitely strange

File: src/views/Login.vue [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+// Libraries
+import axios from 'axios'
+
+export default {
+  name: 'Login.vue',
+
+  components: { ArrowCircleRight },
+
+  data() {
```
Comment (1whatleytay) [inline src/views/Login.vue:64]: Vue3 one day 😢

File: src/views/Login.vue [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+      error: null
+    }
+  },
+
+  methods: {
+    async login() {
+      this.error = null
+
+      axios
```
Comment (1whatleytay) [inline src/views/Login.vue:77]: The login method is `async`, so I'd recommend using `try`/`catch` and `await` instead of `then` and `catch`. It's a bit hard to follow as a chain of events IMO and it's a bit less verbose with `await`.

File: src/views/Login.vue [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        })
+        .then(result => {
+          localStorage.setItem('auth:token', result.data)
+          // Navigate to home
+          window.location = '/'
+        })
+        .catch(error => {
+          if (error.response) {
+            // If the response was Unauthorised
+            if (error.response.status == 401) {
```
Comment (1whatleytay) [inline src/views/Login.vue]:
```suggestion
            if (error.response.status === 401) {
```
Don't take risks :)

File: src/views/Login.vue [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+              this.error = 'Invalid username or password'
+            } else {
+              this.error = `${error.response.data} (Invalid Token)`
+            }
+
+            console.log(error.response.data)
```
Comment (1whatleytay) [inline src/views/Login.vue]:
```suggestion
            console.error(error.response.data)
```

Why not :)

File: src/views/Login.vue [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+          } else if (error.request) {
+            this.error = `${error.request}`
+          }
```
Comment (1whatleytay) [inline src/views/Login.vue:98]: Erm, I think the request object is a bit more expansive than you might expect. I'm a good 70% confidence level that this branch will:
 1. Always be true.
 2. Dump a bunch of JSON on the user.

Instead of this branch I would `this.error = 'Something went wrong.'`.

**Review by 1whatleytay — CHANGES_REQUESTED** [review]
Another more helpful review :)

Hope this helps.

File: src/views/Login.vue (hunk 7 above)
Comment (mk5med) [reply src/views/Login.vue:77]: Personally, the formatting and spacing of logic within the `try/catch` block is quite spread out, the Promise pattern makes the case handling clear and formats it into tight blocks that can be read linearly.

File: src/views/Login.vue (hunk 10 above)
Comment (mk5med) [reply src/views/Login.vue:98]: From my testing the expected errors for this section were caught by the vue proxy and sent in `error.response` (proxy failure). I am not familiar with which other errors are caught by vue proxy, but in cases where this method runs it would be helpful for us to have access to that information so we can take action.

Comment (1whatleytay) [reply src/views/Login.vue:98]: Proxy errors? Yikes.

Sure we can leave this in.

File: src/views/Login.vue (hunk 7 above)
Comment (1whatleytay) [reply src/views/Login.vue:77]: I used to like the `then`/`catch` syntax, but I've come around to `async`/`await`. If you want to keep this though, it would be smart to drop `async` to avoid confusion.

`async`/`await` also has some nice perpetuating benefits done for you. For example, if another function were to call `login` and then would like to do something when the login completes, they're out of luck with `then`/`catch` but can easily queue up events with a function that uses `async`/`await`.
